producer-c/producer-cloudwatch-integ/CanaryStreamCallbacks.cpp

```diff
@@ -0,0 +1,216 @@
+    UINT64 time = GETTIME();
+    UINT64 timeOfFragmentEndSent = pCanaryStreamCallbacks->timeOfNextKeyFrame->find(ack->timestamp)->second;
+    if (ack->ackType != FRAGMENT_ACK_TYPE_BUFFERING) {
+        DLOGD("ack type %u latency %" PRIu64, ack->ackType, (GETTIME() - timeOfFragmentEndSent) / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+    }
+    switch (ack->ackType) {
+        case FRAGMENT_ACK_TYPE_RECEIVED:
+            pCanaryStreamCallbacks->receivedAckDatum.SetValue((GETTIME() - timeOfFragmentEndSent) / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+            pCanaryStreamCallbacks->receivedAckDatum.SetUnit(Aws::CloudWatch::Model::StandardUnit::Milliseconds);
+            CanaryStreamSendMetrics(pCanaryStreamCallbacks, pCanaryStreamCallbacks->receivedAckDatum);
+            break;
+        case FRAGMENT_ACK_TYPE_PERSISTED:
+            pCanaryStreamCallbacks->persistedAckDatum.SetValue((GETTIME() - timeOfFragmentEndSent) / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+            pCanaryStreamCallbacks->persistedAckDatum.SetUnit(Aws::CloudWatch::Model::StandardUnit::Milliseconds);
+            CanaryStreamSendMetrics(pCanaryStreamCallbacks, pCanaryStreamCallbacks->persistedAckDatum);
+            pCanaryStreamCallbacks->timeOfNextKeyFrame->erase(ack->timestamp);
+            break;
+        case FRAGMENT_ACK_TYPE_ERROR:
+            DLOGE("Received Error Ack timestamp %" PRIu64 " fragment number %s error code %lu", ack->timestamp, ack->sequenceNumber, ack->result);
+            break;
+        default:
+            break;
+    }
+    DLOGD("CanaryStreamFragmentAckHandler took %" PRIu64, GETTIME() - time);
+    return STATUS_SUCCESS;
+}
+
+VOID OnPutMetricDataResponseReceivedHandler(const Aws::CloudWatch::CloudWatchClient* cwClient,
+                                            const Aws::CloudWatch::Model::PutMetricDataRequest& request,
+                                            const Aws::CloudWatch::Model::PutMetricDataOutcome& outcome,
+                                            const std::shared_ptr<const Aws::Client::AsyncCallerContext>& context) {
+    if (!outcome.IsSuccess())
+    {
+        DLOGE("Failed to put sample metric data: %s", outcome.GetError().GetMessage().c_str());
+    }
+    else
+    {
+        DLOGV("Successfully put sample metric data");
+    }
+}
+
+VOID CanaryStreamSendMetrics(PCanaryStreamCallbacks pCanaryStreamCallbacks, Aws::CloudWatch::Model::MetricDatum& metricDrum)
+{
+    Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+    cwRequest.SetNamespace("KinesisVideoSDKCanary");
+    cwRequest.AddMetricData(metricDrum);
+    pCanaryStreamCallbacks->pCwClient->PutMetricDataAsync(cwRequest, OnPutMetricDataResponseReceivedHandler);
+}
+
+VOID ComputeStreamMetricsFromCanary(STREAM_HANDLE streamHandle, PCanaryStreamCallbacks pCanaryStreamCallbacks) {
+    StreamMetrics canaryStreamMetrics;
+    canaryStreamMetrics.version = STREAM_METRICS_CURRENT_VERSION;
+    getKinesisVideoStreamMetrics(streamHandle, &canaryStreamMetrics);
```

Comment:
check the result. If not successful then you will end up with garbage.

Reply:
Done